File: gitlfs-client/src/main/java/ru/bozaro/gitlfs/client/auth/BasicAuthProvider.java
package ru.bozaro.gitlfs.client.auth;

import org.apache.commons.codec.binary.Base64;
import ru.bozaro.gitlfs.common.data.Link;
import ru.bozaro.gitlfs.common.data.Operation;

import javax.annotation.CheckForNull;
import javax.annotation.Nonnull;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.TreeMap;

import static ru.bozaro.gitlfs.common.Constants.HEADER_AUTHORIZATION;

/**
 * Auth provider for basic authentication.
 *
 * @author Artem V. Navrotskiy
 */
public class BasicAuthProvider implements AuthProvider {
  @Nonnull
  private final Link auth;

  public BasicAuthProvider(@Nonnull URI href) {
    this(href, null, null);
  }

  public BasicAuthProvider(@Nonnull URI href, @CheckForNull String login, @CheckForNull String password) {
    final String authLogin;
    if (isEmpty(login)) {
      authLogin = extractLogin(href.getUserInfo());
    } else {
      authLogin = login;
    }
    final String authPassword;
    if (isEmpty(password)) {
      authPassword = extractPassword(href.getUserInfo());
    } else {
      authPassword = password;
    }
    final TreeMap<String, String> header = new TreeMap<>();
    final String userInfo = authLogin + ':' + authPassword;
    header.put(HEADER_AUTHORIZATION, "Basic " + new String(Base64.encodeBase64(userInfo.getBytes(StandardCharsets.UTF_8)), StandardCharsets.UTF_8));
    try {
      final String scheme = "git".equals(href.getScheme()) ? "https" : href.getScheme();
      this.auth = new Link(new URI(scheme, href.getAuthority(), href.getPath(), null, null), Collections.unmodifiableMap(header), null);
    } catch (URISyntaxException e) {
      throw new IllegalStateException(e);
    }
  }

  private static boolean isEmpty(@CheckForNull String value) {
    return value == null || value.isEmpty();
  }

  @Nonnull
  private static String extractLogin(@CheckForNull String userInfo) {
    if (userInfo == null) return "";
    final int separator = userInfo.indexOf(':');
    return (separator >= 0) ? userInfo.substring(0, separator) : userInfo;
  }

  @Nonnull
  private static String extractPassword(@CheckForNull String userInfo) {
    if (userInfo == null) return "";
    final int separator = userInfo.indexOf(':');
    return (separator >= 0) ? userInfo.substring(separator + 1) : "";
  }

  @Nonnull
  @Override
  public Link getAuth(@Nonnull Operation operation) {
    return auth;
  }

  @Override
  public void invalidateAuth(@Nonnull Operation operation, @Nonnull Link auth) {
  }
}

File: gitlfs-client/src/main/kotlin/ru/bozaro/gitlfs/client/auth/BasicAuthProvider.kt
package ru.bozaro.gitlfs.client.auth

import org.apache.commons.codec.binary.Base64
import ru.bozaro.gitlfs.common.Constants.HEADER_AUTHORIZATION
import ru.bozaro.gitlfs.common.data.Link
import ru.bozaro.gitlfs.common.data.Operation
import java.net.URI
import java.net.URISyntaxException
import java.nio.charset.StandardCharsets
import java.util.*

/**
 * Auth provider for basic authentication.
 *
 * @author Artem V. Navrotskiy
 */
class BasicAuthProvider constructor(href: URI, login: String? = null, password: String? = null) : AuthProvider {
    private var auth: Link

    override fun getAuth(operation: Operation): Link {
        return auth
    }

    override fun invalidateAuth(operation: Operation, auth: Link) {}

    companion object {
        private fun isEmpty(value: String?): Boolean {
            return value == null || value.isEmpty()
        }

        private fun extractLogin(userInfo: String?): String {
            if (userInfo == null) return ""
            val separator = userInfo.indexOf(':')
            return if (separator >= 0) userInfo.substring(0, separator) else userInfo
        }

        private fun extractPassword(userInfo: String?): String {
            if (userInfo == null) return ""
            val separator = userInfo.indexOf(':')
            return if (separator >= 0) userInfo.substring(separator + 1) else ""
        }
    }

    init {
        val authLogin: String? = if (isEmpty(login)) {
            extractLogin(href.userInfo)
        } else {
            login
        }
        val authPassword: String? = if (isEmpty(password)) {
            extractPassword(href.userInfo)
        } else {
            password
        }
        val header = TreeMap<String, String>()
        val userInfo = "$authLogin:$authPassword"
        header[HEADER_AUTHORIZATION] =
                "Basic " + String(Base64.encodeBase64(userInfo.toByteArray(StandardCharsets.UTF_8)), StandardCharsets.UTF_8)
        try {
            val scheme = if ("git" == href.scheme) "https" else href.scheme
            auth = Link(URI(scheme, href.authority, href.path, null, null), Collections.unmodifiableMap(header), null)
        } catch (e: URISyntaxException) {
            throw IllegalStateException(e)
        }
    }
}
